File: tests/python/google-genai/test_prototype.py
"""Conformance test: prototype instrumentation for Google GenAI.

Exercises: chat, chat_streaming, embeddings
against a mock Google GenAI server, with manual OTel spans.
"""

import json
import os

from otel_setup import setup_otel, flush_and_shutdown
from opentelemetry import trace
from opentelemetry._logs import get_logger_provider

MOCK_BASE_URL = os.environ["MOCK_LLM_URL"]

_prototype_tracer = trace.get_tracer("gen_ai.prototype")


def run_chat():
    """Scenario: basic chat completion with prototype instrumentation."""
    from google import genai
    from google.genai import types

    print("  [chat] basic chat completion via Google GenAI (prototype)")
    client = genai.Client(
        api_key="mock-key",
        http_options=types.HttpOptions(
            base_url=MOCK_BASE_URL,
            api_version="v1beta",
        ),
    )
    request_model = "gemini-2.0-flash"
    with _prototype_tracer.start_as_current_span("chat gemini-2.0-flash") as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "google_genai")
        span.set_attribute("gen_ai.request.model", request_model)
        prompt_text = "Say hello."
        response = client.models.generate_content(
            model=request_model,
            contents=prompt_text,
        )
        if response.model_version:
            span.set_attribute("gen_ai.response.model", response.model_version)
        if response.candidates and response.candidates[0].finish_reason:
            span.set_attribute("gen_ai.response.finish_reasons", [str(response.candidates[0].finish_reason)])
        if hasattr(response, "usage_metadata") and response.usage_metadata:
            if hasattr(response.usage_metadata, "prompt_token_count") and response.usage_metadata.prompt_token_count:
                span.set_attribute("gen_ai.usage.input_tokens", response.usage_metadata.prompt_token_count)
            if hasattr(response.usage_metadata, "candidates_token_count") and response.usage_metadata.candidates_token_count:
                span.set_attribute("gen_ai.usage.output_tokens", response.usage_metadata.candidates_token_count)
            if hasattr(response.usage_metadata, "thoughts_token_count") and response.usage_metadata.thoughts_token_count:
                span.set_attribute("gen_ai.usage.reasoning.output_tokens", response.usage_metadata.thoughts_token_count)

        # Emit inference operation details event
        event_attrs = {
            "gen_ai.operation.name": "chat",
            "gen_ai.request.model": request_model,
            "gen_ai.input.messages": json.dumps([
                {"role": "user", "parts": [{"type": "text", "content": prompt_text}]}
            ]),
            "gen_ai.output.messages": json.dumps([
                {
                    "role": "assistant",
                    "parts": [{"type": "text", "content": response.text}],
                    "finish_reason": str(response.candidates[0].finish_reason) if response.candidates else None,
                }
            ]),
        }
        if response.model_version:
            event_attrs["gen_ai.response.model"] = response.model_version
        if response.candidates and response.candidates[0].finish_reason:
            event_attrs["gen_ai.response.finish_reasons"] = [str(response.candidates[0].finish_reason)]
        if hasattr(response, "usage_metadata") and response.usage_metadata:
            if hasattr(response.usage_metadata, "prompt_token_count") and response.usage_metadata.prompt_token_count:
                event_attrs["gen_ai.usage.input_tokens"] = response.usage_metadata.prompt_token_count
            if hasattr(response.usage_metadata, "candidates_token_count") and response.usage_metadata.candidates_token_count:
                event_attrs["gen_ai.usage.output_tokens"] = response.usage_metadata.candidates_token_count
            if hasattr(response.usage_metadata, "thoughts_token_count") and response.usage_metadata.thoughts_token_count:
                event_attrs["gen_ai.usage.reasoning.output_tokens"] = response.usage_metadata.thoughts_token_count
        get_logger_provider().get_logger("gen_ai.prototype").emit(
            event_name="gen_ai.client.inference.operation.details",
            body="Inference operation details",
            attributes=event_attrs,
        )

        print(f"    -> {response.text[:60]}")


def run_chat_tool_call():
    """Scenario: chat with tool calling with prototype instrumentation."""
    from google import genai
    from google.genai import types

    print("  [chat_tool_call] chat with tool calling via Google GenAI (prototype)")
    client = genai.Client(
        api_key="mock-key",
        http_options=types.HttpOptions(
            base_url=MOCK_BASE_URL,
            api_version="v1beta",
        ),
    )
    request_model = "gemini-2.0-flash"
    tool = types.Tool(function_declarations=[
        types.FunctionDeclaration(
            name="get_weather",
            description="Get the current weather",
            parameters=types.Schema(
                type="OBJECT",
                properties={
                    "location": types.Schema(type="STRING", description="City name"),
                },
                required=["location"],
            ),
        )
    ])
    tools = [tool]
    with _prototype_tracer.start_as_current_span("chat gemini-2.0-flash") as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "google_genai")
        span.set_attribute("gen_ai.request.model", request_model)
        span.set_attribute("gen_ai.tool.definitions", json.dumps([{
            "function_declarations": [{
                "name": "get_weather",
                "description": "Get the current weather",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "location": {"type": "string", "description": "City name"},
                    },
                    "required": ["location"],
                },
            }]
        }]))
        response = client.models.generate_content(
            model=request_model,
            contents="What's the weather in Seattle?",
            config=types.GenerateContentConfig(tools=tools),
        )
        if response.model_version:
            span.set_attribute("gen_ai.response.model", response.model_version)
        if response.candidates and response.candidates[0].finish_reason:
            span.set_attribute("gen_ai.response.finish_reasons", [str(response.candidates[0].finish_reason)])
        if hasattr(response, "usage_metadata") and response.usage_metadata:
            if hasattr(response.usage_metadata, "prompt_token_count") and response.usage_metadata.prompt_token_count:
                span.set_attribute("gen_ai.usage.input_tokens", response.usage_metadata.prompt_token_count)
            if hasattr(response.usage_metadata, "candidates_token_count") and response.usage_metadata.candidates_token_count:
                span.set_attribute("gen_ai.usage.output_tokens", response.usage_metadata.candidates_token_count)
        if response.candidates and response.candidates[0].content.parts:
            part = response.candidates[0].content.parts[0]
            if hasattr(part, "function_call") and part.function_call:
                print(f"    -> tool_call: {part.function_call.name}")
            else:
                print(f"    -> {response.text[:60]}")
        else:
            print(f"    -> {response.text[:60]}")


def run_chat_streaming():
    """Scenario: streaming chat completion with prototype instrumentation."""
    from google import genai
    from google.genai import types

    print("  [chat_streaming] streaming chat via Google GenAI (prototype)")
    client = genai.Client(
        api_key="mock-key",
        http_options=types.HttpOptions(
            base_url=MOCK_BASE_URL,
            api_version="v1beta",
        ),
    )
    request_model = "gemini-2.0-flash"
    with _prototype_tracer.start_as_current_span("chat gemini-2.0-flash") as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "google_genai")
        span.set_attribute("gen_ai.request.model", request_model)
        text = ""
        last_chunk = None
        for chunk in client.models.generate_content_stream(
            model=request_model,
            contents="Tell me a joke.",
        ):
            if chunk.text:
                text += chunk.text
            last_chunk = chunk
        if last_chunk and last_chunk.model_version:
            span.set_attribute("gen_ai.response.model", last_chunk.model_version)
        if last_chunk and last_chunk.candidates and last_chunk.candidates[0].finish_reason:
            span.set_attribute("gen_ai.response.finish_reasons", [str(last_chunk.candidates[0].finish_reason)])
        if last_chunk and hasattr(last_chunk, "usage_metadata") and last_chunk.usage_metadata:
            if hasattr(last_chunk.usage_metadata, "prompt_token_count") and last_chunk.usage_metadata.prompt_token_count:
                span.set_attribute("gen_ai.usage.input_tokens", last_chunk.usage_metadata.prompt_token_count)
            if hasattr(last_chunk.usage_metadata, "candidates_token_count") and last_chunk.usage_metadata.candidates_token_count:
                span.set_attribute("gen_ai.usage.output_tokens", last_chunk.usage_metadata.candidates_token_count)
            if hasattr(last_chunk.usage_metadata, "thoughts_token_count") and last_chunk.usage_metadata.thoughts_token_count:
                span.set_attribute("gen_ai.usage.reasoning.output_tokens", last_chunk.usage_metadata.thoughts_token_count)
        print(f"    -> {text[:60]}")


def run_embeddings():
    """Scenario: embedding generation with prototype instrumentation."""
    from google import genai
    from google.genai import types

    print("  [embeddings] embedding generation via Google GenAI (prototype)")
    client = genai.Client(
        api_key="mock-key",
        http_options=types.HttpOptions(
            base_url=MOCK_BASE_URL,
            api_version="v1beta",
        ),
    )
    request_model = "text-embedding-004"
    with _prototype_tracer.start_as_current_span("embeddings text-embedding-004") as span:
        span.set_attribute("gen_ai.operation.name", "embeddings")
        span.set_attribute("gen_ai.provider.name", "google_genai")
        span.set_attribute("gen_ai.request.model", request_model)
        response = client.models.embed_content(
            model=request_model,
            contents="Hello, world!",
        )
        print(f"    -> embedding dim: {len(response.embeddings[0].values)}")


def main():
    print("=== Prototype: Google GenAI Conformance Test ===")

    tp, lp, mp = setup_otel()
    # NO instrument() call – prototype instrumentation only

    run_chat()
    run_chat_tool_call()
    run_chat_streaming()
    run_embeddings()

    flush_and_shutdown(tp, lp, mp)


if __name__ == "__main__":
    main()

File: tests/js/vertexai/test_prototype.ts
/**
 * Conformance test: Prototype instrumentation for Vertex AI.
 */
import { trace } from "@opentelemetry/api";
import { logs, SeverityNumber } from "@opentelemetry/api-logs";
import { flushAndShutdownOtel, setupOtel } from "../otel";

const nodeProcess = globalThis.process as NodeJS.Process & {
  env: Record<string, string | undefined>;
  exit(code?: number): never;
};

const MOCK_LLM_URL = nodeProcess.env.MOCK_LLM_URL!;
const tracer = trace.getTracer("gen_ai.prototype");

// Patch fetch before any SDK imports to intercept auth and redirect HTTPS→HTTP.
const _originalFetch = globalThis.fetch;
globalThis.fetch = async function (input: any, init?: any) {
  let url =
    typeof input === "string"
      ? input
      : input instanceof URL
        ? input.href
        : input.url;

  // Mock OAuth token endpoint so the SDK gets a fake access token
  if (url.includes("oauth2.googleapis.com") || url.includes("/token")) {
    return new Response(
      JSON.stringify({
        access_token: "mock-token",
        token_type: "Bearer",
        expires_in: 3600,
      }),
      { status: 200, headers: { "content-type": "application/json" } }
    );
  }

  // Redirect HTTPS to HTTP for mock server
  url = url.replace(/^https:\/\//, "http://");
  return _originalFetch(url, init);
};

async function main() {
  console.log("=== Prototype: Vertex AI JS Conformance Test ===");
  const otel = setupOtel();

  const { VertexAI } = await import("@google-cloud/vertexai");
  const { createFakeGoogleAuth } = await import(
    "@google-cloud/vertexai/build/src/testing/fake_google_auth"
  );
  const mockUrl = new URL(MOCK_LLM_URL);
  const vertexai = new VertexAI({
    project: "test-project",
    location: "us-central1",
    apiEndpoint: `${mockUrl.hostname}:${mockUrl.port}`,
  });
  (vertexai as any).googleAuth = createFakeGoogleAuth({
    accessToken: "mock-token",
  });
  const requestModel = "gemini-2.0-flash";
  const model = vertexai.getGenerativeModel({ model: requestModel });

  // Scenario: chat
  console.log("  [chat] basic content generation (prototype)");
  await tracer.startActiveSpan("chat gemini-2.0-flash", async (span) => {
    span.setAttribute("gen_ai.operation.name", "chat");
    span.setAttribute("gen_ai.provider.name", "vertex_ai");
    span.setAttribute("gen_ai.request.model", requestModel);
    const userMessage = "Say hello.";
    const result = await model.generateContent(userMessage);
    const resp = result.response;
    const responseModel = (resp as any).modelVersion ?? requestModel;
    span.setAttribute("gen_ai.response.model", responseModel);
    const candidate = resp.candidates?.[0];
    if (candidate?.finishReason) {
      span.setAttribute("gen_ai.response.finish_reasons", [candidate.finishReason]);
    }
    if (resp.usageMetadata) {
      if (resp.usageMetadata.promptTokenCount) {
        span.setAttribute("gen_ai.usage.input_tokens", resp.usageMetadata.promptTokenCount);
      }
      if (resp.usageMetadata.candidatesTokenCount) {
        span.setAttribute("gen_ai.usage.output_tokens", resp.usageMetadata.candidatesTokenCount);
      }
    }
    const text = candidate?.content?.parts?.[0]?.text ?? "";

    // Emit inference operation details event
    logs.getLogger("gen_ai.prototype").emit({
      severityNumber: SeverityNumber.INFO,
      eventName: "gen_ai.client.inference.operation.details",
      body: "Inference operation details",
      attributes: {
        "gen_ai.operation.name": "chat",
        "gen_ai.request.model": requestModel,
        "gen_ai.response.model": responseModel,
        "gen_ai.response.finish_reasons": candidate?.finishReason ? [candidate.finishReason] : undefined,
        "gen_ai.usage.input_tokens": resp.usageMetadata?.promptTokenCount,
        "gen_ai.usage.output_tokens": resp.usageMetadata?.candidatesTokenCount,
        "gen_ai.input.messages": JSON.stringify([
          { role: "user", parts: [{ type: "text", content: userMessage }] }
        ]),
        "gen_ai.output.messages": JSON.stringify([{
          role: "assistant",
          parts: [{ type: "text", content: text }],
          finish_reason: candidate?.finishReason,
        }]),
      },
    });

    console.log(`    -> ${text.slice(0, 60)}`);
    span.end();
  });

  // Scenario: chat with tool calling
  console.log("  [chat_tool_call] content generation with tool calling (prototype)");
  await tracer.startActiveSpan("chat gemini-2.0-flash", async (span) => {
    span.setAttribute("gen_ai.operation.name", "chat");
    span.setAttribute("gen_ai.provider.name", "vertex_ai");
    span.setAttribute("gen_ai.request.model", requestModel);
    const requestTool = {
      functionDeclarations: [{
        name: "get_weather",
        description: "Get the current weather",
        parameters: {
          type: "OBJECT" as const,
          properties: {
            location: { type: "STRING" as const, description: "City name" },
          },
          required: ["location"],
        },
      }],
    };
    span.setAttribute("gen_ai.tool.definitions", JSON.stringify([requestTool]));
    const result = await model.generateContent({
      contents: [{ role: "user", parts: [{ text: "What's the weather in Seattle?" }] }],
      tools: [requestTool],
    });
    const resp = result.response;
    const responseModel = (resp as any).modelVersion ?? requestModel;
    span.setAttribute("gen_ai.response.model", responseModel);
    const candidate = resp.candidates?.[0];
    if (candidate?.finishReason) {
      span.setAttribute("gen_ai.response.finish_reasons", [candidate.finishReason]);
    }
    if (resp.usageMetadata) {
      if (resp.usageMetadata.promptTokenCount) {
        span.setAttribute("gen_ai.usage.input_tokens", resp.usageMetadata.promptTokenCount);
      }
      if (resp.usageMetadata.candidatesTokenCount) {
        span.setAttribute("gen_ai.usage.output_tokens", resp.usageMetadata.candidatesTokenCount);
      }
    }
    const part = candidate?.content?.parts?.[0];
    if (part && "functionCall" in part && part.functionCall) {
      console.log(`    -> tool_call: ${part.functionCall.name}`);
    } else {
      const text = part?.text ?? "";
      console.log(`    -> ${text.slice(0, 60)}`);
    }
    span.end();
  });

  // Scenario: streaming chat
  console.log("  [chat_streaming] streaming content generation (prototype)");
  await tracer.startActiveSpan("chat gemini-2.0-flash", async (span) => {
    span.setAttribute("gen_ai.operation.name", "chat");
    span.setAttribute("gen_ai.provider.name", "vertex_ai");
    span.setAttribute("gen_ai.request.model", requestModel);
    const result = await model.generateContentStream("Tell me a joke.");
    let text = "";
    for await (const chunk of result.stream) {
      const part = chunk.candidates?.[0]?.content?.parts?.[0]?.text;
      if (part) text += part;
    }
    const aggregated = await result.response;
    const responseModel = (aggregated as any).modelVersion ?? requestModel;
    span.setAttribute("gen_ai.response.model", responseModel);
    const candidate = aggregated.candidates?.[0];
    if (candidate?.finishReason) {
      span.setAttribute("gen_ai.response.finish_reasons", [candidate.finishReason]);
    }
    if (aggregated.usageMetadata) {
      if (aggregated.usageMetadata.promptTokenCount) {
        span.setAttribute("gen_ai.usage.input_tokens", aggregated.usageMetadata.promptTokenCount);
      }
      if (aggregated.usageMetadata.candidatesTokenCount) {
        span.setAttribute("gen_ai.usage.output_tokens", aggregated.usageMetadata.candidatesTokenCount);
      }
    }
    console.log(`    -> ${text.slice(0, 60)}`);
    span.end();
  });

  await flushAndShutdownOtel(otel);
}

main().catch((e) => { console.error(e); nodeProcess.exit(1); });
